marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+```haskell
+fromBuiltinData datum          ≡ (Just marloweData   :: Maybe MarloweData)
+fromBuiltinData redeemer       ≡ (Just markloweInput :: Maybe MarloweInput)
+fromBuiltinData scriptContext' ≡ (Just scriptContext :: Maybe ScriptContext)
+```
+
+
+#### *Constraint 2.* Single Marlowe script input
+
+The output of exactly one Marlowe UTxO is spent in the transaction.
+```haskell
+filter (isMaroweAddress . txOutAddress . txInfoResolved) (txInfoInputs $ scriptContextTxInfo scriptContext) ≡ [TxInInfo _ (TxOut _ inValue _)]
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+Hereafter we use `inValue` as the script value being spent.
+
+
+#### *Constraint 3.* Single Marlowe output
+
+There is a single output to the Marlowe script address unless the contract closes.
+```haskell
+txOutContract transactionOutput /= Close ⇒ filter (isMaroweAddress . txOutAddress) (txInfoOutputs $ scriptContextTxInfo scriptContext) ≡ [TxOut _ outValue (Just outDatum)]
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+Hereafter we use `outValue` for the value being output to the script address and `outDatum` being the datum associated with that UTxO.
+
+
+#### *Constraint 4.* No output to script on close
+
+When a contract closes, the transaction does not output to the Marlowe script address.
+```haskell
+txOutContract transactionOutput == Close ⇒ filter (isMaroweAddress . txOutAddress) (txInfoOutputs $ scriptContextTxInfo scriptContext) ≡ []
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+
+
+#### *Constraint 5.* Input value from script
+
+The value of the Marlowe UTxO being spent exactly matches the value recorded in the Marlowe state.
+```haskell
+inValue ≡ foldMap toValue (accounts $ marloweState marloweData)
+  where toValue ((_, Token currency name), count)) = Value.singleton currency name count
+```
+
+
+#### *Constraint 6.* Output value to script
+
+The beginning balance plus the deposits equals the ending balance plus the payments.
+```haskell
+inValue + foldMap valueOfDeposit (fmap getInputContent transactionInput) ≡ outValue + foldMap valueOfPayment (txOutPayments transactionOutput)
+  where getInputContent (Input inputContent) = inputContent
+        getInputContent (MerkleizedTxInput inputContent _) = inputContent
+        valueOfDeposit (IDeposit _ _ (Token currency name) count) = Value.singleton currency name count
+        valueOfDeposit _ = mempty
+        valueOfPayment (Payment _ (Party _) value)) = value
+        valueOfPayment _ = mempty
+```
+
+
+#### *Constraint 7.* Input state
+
+The input state of the transaction is used in the semantics computation.
```

Comment:
Needs rephrasing: doesn't seem to express the force of the constraint. Maybe "The input state of the transaction is the state used in the semantics computation."?

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0